### PR TITLE
Fix rejected identity accept encoding

### PR DESCRIPTION
**Pull request by colega**
#### What this PR does

Fixes response not being gzip-compressed when identity is explicitly rejected by the `Accept-Encoding: identity;q=0` header.

This PR brings the code from https://github.com/nytimes/gziphandler/pull/112 into `pkg/util/gziphandler` since it's unlikely for that PR to be merged soon.

I'd also consider switching to some lighter implementation, I haven't deeply checked but Prometheus uses [much less code](https://github.com/prometheus/prometheus/blob/af0f6da5cb5bb9736abcfc9e8c7633ee01000ce2/util/httputil/compression.go) for a similar purpose (although they lack qvalue check)

#### Which issue(s) this PR fixes or relates to

Fixes https://github.com/grafana/mimir/issues/1855

#### Checklist

- [x] Tests updated
- [ ] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

**Comment by colega** [comment]
This is how I fixed this in Prometheus: https://github.com/prometheus/prometheus/pull/10706
That's much less code for the same purpose, except that it does not play with content-types and it encodes the response even if it's already comes encoded (that's better for wrapped middlewares). I wonder if we'd prefer to use that one instead.

Maybe we could consider that in a separate PR.

File: pkg/api/api_test.go [hunk 1]
```diff
@@ -62,3 +69,144 @@ func TestNewApiWithInvalidSourceIPExtractor(t *testing.T) {
 	require.Error(t, err)
 	require.Nil(t, api)
 }
+
+func TestApiGzip(t *testing.T) {
+	cfg := Config{}
+	serverCfg := getServerConfig(t)
+	srv, err := server.New(serverCfg)
+	require.NoError(t, err)
+	go func() { _ = srv.Run() }()
```
Comment (johannaratliff) [inline pkg/api/api_test.go:78]: `go srv.Run()` here is equivalent

File: pkg/util/gziphandler/gzip.go [hunk 2]
```diff
@@ -81,19 +86,12 @@ type GzipResponseWriter struct {
...
-}
-
-type GzipResponseWriterWithCloseNotify struct {
-	*GzipResponseWriter
-}
-
-func (w GzipResponseWriterWithCloseNotify) CloseNotify() <-chan bool {
-	return w.ResponseWriter.(http.CloseNotifier).CloseNotify()
+	contentTypes    []parsedContentType // Only compress if the response is one of these content-types. All are accepted if empty.
+	acceptsIdentity bool                // If false, then request explicitly rejected non-encoded requests.
```
Comment (johannaratliff) [inline pkg/util/gziphandler/gzip.go:94]: The comment here seems like a double negative and I'm confused. If `acceptsIdentity` is false, then we reject non-encoded requests? The extra phrasing in the comment is throwing me off

File: pkg/util/gziphandler/gzip.go [hunk 3]
```diff
@@ -118,19 +116,21 @@ func (w *GzipResponseWriter) Write(b []byte) (int, error) {
 		ce    = w.Header().Get(contentEncoding)
 	)
 	// Only continue if they didn't already choose an encoding or a known unhandled content length or type.
-	if ce == "" && (cl == 0 || cl >= w.minSize) && (ct == "" || handleContentType(w.contentTypes, ct)) {
+	if ce == "" && (cl == 0 || cl >= w.minSize || !w.acceptsIdentity) && (ct == "" || handleContentType(w.contentTypes, ct)) {
```
Comment (johannaratliff) [inline pkg/util/gziphandler/gzip.go:119]: I know you just appended here, but the double negatives in these comments makes intent hard to follow

File: pkg/util/gziphandler/gzip.go [hunk 4]
```diff
@@ -393,15 +391,15 @@ func (c *config) validate() error {
 	return nil
 }
```
Comment (johannaratliff) [inline pkg/util/gziphandler/gzip.go:393]: This is a nit, but with option patterns, I really like the `WithMyOption` naming pattern (here `WithCompressionLevel` so that it shows the additive nature of running the func on the config). I figure since we're taking the code anyway, might as well adjust for readability and clarity.

**Review by johannaratliff — APPROVED** [review]
These are all nits that I think would help with clarity, but the PR logic itself LGTM

File: pkg/util/gziphandler/gzip.go (hunk 4 above)
Comment (johannaratliff) [reply pkg/util/gziphandler/gzip.go:393]: I see now that it clarifies by using `GzipHandlerWithOps(opts ...Option)` so this comment is redundant

File: pkg/util/gziphandler/gzip.go (hunk 3 above)
Comment (colega) [reply pkg/util/gziphandler/gzip.go:119]: I completely agree, but I wanted to keep the code changes at minimum compared to the original code. Maybe I should fix that code.

File: pkg/api/api_test.go (hunk 1 above)
Comment (colega) [reply pkg/api/api_test.go:78]: But the linter complained :(

File: pkg/util/gziphandler/gzip.go [hunk 5]
```diff
@@ -445,11 +443,25 @@ func GzipHandler(h http.Handler) http.Handler {
...
 
-// acceptsGzip returns true if the given HTTP request indicates that it will
-// accept a gzipped response.
-func acceptsGzip(r *http.Request) bool {
+// requestAcceptance checks whether a given HTTP request indicates that it will
+// accept a gzipped response and whether it's going to reject an non-encoded response.
+//
+// acceptsGzip is true if the given HTTP request indicates that it will
+// accept a gzipped response and/or an identity request.
+// acceptsIdentity is if the given HTTP request didn't explicitly exclude identity encoding.
```
Comment (pracucci) [inline pkg/util/gziphandler/gzip.go:451]:
```suggestion
// acceptsIdentity is true if the given HTTP request didn't explicitly exclude identity encoding.
```

**Review by pracucci — COMMENTED** [review]
Have you considered vendoring your fork, instead of copying code into Mimir? Also, if we copy code here, then I would add a faillint rule (in Makefile) to ensure we don't use the library anymore.

File: pkg/util/gziphandler/gzip.go [hunk 6]
```diff
@@ -322,21 +324,17 @@ func GzipHandlerWithOpts(opts ...option) (func(http.Handler) http.Handler, error
...
-					contentTypes:   c.contentTypes,
+					ResponseWriter:  w,
+					index:           index,
+					minSize:         c.minSize,
+					contentTypes:    c.contentTypes,
+					acceptsIdentity: acceptsIdentity,
 				}
 				defer gw.Close()
 
-				if _, ok := w.(http.CloseNotifier); ok {
```
[the remark this answers is not in the document]
Comment (colega) [reply pkg/util/gziphandler/gzip.go:334]: From the `http.CloseNotifier` comment:

```go
// Deprecated: the CloseNotifier interface predates Go's context package.
// New code should use Request.Context instead.
```

This was here just because the source library has been unmaintained for years, but when I brought the code here, linter complained about it.

**Comment by colega** [comment]
Added the original package to faillint. I don't want to import my fork (since then I'd need either to change my fork's module name, or we'd need to take care of the replace in enterprise metrics too), I think that having the code here is just easier.

**Review by pracucci — APPROVED** [review]
Changes make sense to me. Logic is non trivial, but nothing wrong jumped to my 👀

File: pkg/api/api_test.go (hunk 1 above)
Comment (johannaratliff) [reply pkg/api/api_test.go:78]: That's.... weird. Is it just because it didn't like the lack of return value? I feel like `go srv.Run()` is better from a reading perspective for anyone who doesn't know annoying go quirks

File: pkg/util/gziphandler/gzip.go (hunk 3 above)
Comment (colega) [reply pkg/util/gziphandler/gzip.go]: I finally refactored this, trying to make it cleaner. I did it in multiple commits to make the reviewing (hopefully) easier, PTAL:
- https://github.com/grafana/mimir/pull/1864/commits/d27a4eb59956ce08e5db09349e3532621ef3b461
- https://github.com/grafana/mimir/pull/1864/commits/a8673b307c87f6aa67280a61682fc8355f26243a
- https://github.com/grafana/mimir/pull/1864/commits/d22899c1c8d01fda74278a48db653bd050dcf669
- https://github.com/grafana/mimir/pull/1864/commits/705540dcb40c16de631b778770a53ccc6b079291
- https://github.com/grafana/mimir/pull/1864/commits/a21356223b79d9041c66b9049540f09768adab27

File: pkg/util/gziphandler/gzip.go (hunk 2 above)
Comment (colega) [reply pkg/util/gziphandler/gzip.go]: Hopefully it's better now since https://github.com/grafana/mimir/pull/1864/commits/705540dcb40c16de631b778770a53ccc6b079291 as part of the refactor I mentioned below.

File: pkg/util/gziphandler/gzip.go (hunk 3 above)
Comment (colega) [reply pkg/util/gziphandler/gzip.go]: @johannaratliff I'll merge this as it is now, but I'm open to any extra feedback you may have.
